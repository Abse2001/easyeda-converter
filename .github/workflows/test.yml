name: Test and Benchmark

on:
  pull_request:
    branches:
      - main

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v2

      - name: Use bun
        uses: oven-sh/setup-bun@v1

      - name: Install dependencies
        run: bun ci

      - name: Pull kicad footprints
        run: bun test
