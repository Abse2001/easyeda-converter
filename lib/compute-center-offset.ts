import type { z } from "zod"
import type { EasyEdaJson } from "./schemas/easy-eda-json-schema"
import type { PadSchema } from "./schemas/package-detail-shape-schema"
import { mm } from "@tscircuit/mm"

export const computeCenterOffset = (easyeda: EasyEdaJson) => {
  const pads = easyeda.packageDetail.dataStr.shape.filter(
    (shape): shape is z.infer<typeof PadSchema> => shape.type === "PAD"
  )

  const minX = Math.min(...pads.map((pad) => mm(pad.center.x)))
  const maxX = Math.max(...pads.map((pad) => mm(pad.center.x)))
  const minY = Math.min(...pads.map((pad) => mm(pad.center.y)))
  const maxY = Math.max(...pads.map((pad) => mm(pad.center.y)))

  const centerX = (minX + maxX) / 2
  const centerY = (minY + maxY) / 2

  return {
    x: centerX,
    y: centerY,
  }
}
