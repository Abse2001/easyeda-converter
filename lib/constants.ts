export const DEFAULT_PCB_THICKNESS_MM = 1.6
