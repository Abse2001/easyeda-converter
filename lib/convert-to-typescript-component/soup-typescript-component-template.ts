import type { AnySoupElement } from "circuit-json"
import type { ChipProps } from "@tscircuit/props"
import { su } from "@tscircuit/soup-util"
import type { BetterEasyEdaJson } from "../schemas/easy-eda-json-schema"
import { generateFootprintTsx } from "../generate-footprint-tsx"

interface Params {
  pinLabels: ChipProps["pinLabels"]
  componentName: string
  schPinArrangement: ChipProps["schPortArrangement"]
  objUrl?: string
  easyEdaJson: BetterEasyEdaJson
}

export const soupTypescriptComponentTemplate = ({
  pinLabels,
  componentName,
  schPinArrangement,
  objUrl,
  easyEdaJson,
}: Params) => {
  const footprintTsx = generateFootprintTsx(easyEdaJson)
  return `
import { createUseComponent } from "@tscircuit/core"
import type { CommonLayoutProps } from "@tscircuit/props"

const pinLabels = ${JSON.stringify(pinLabels, null, "  ")} as const
const pinNames = Object.values(pinLabels)

interface Props extends CommonLayoutProps {
  name: string
}

export const ${componentName} = (props: Props) => {
  return (
    <chip
      {...props}
      footprint={${footprintTsx}}
      ${
        objUrl
          ? `cadModel={{
        objUrl: "${objUrl}"
      }}`
          : ""
      }
      pinLabels={pinLabels}
      schPinSpacing={0.75}
      schPortArrangement={${JSON.stringify(schPinArrangement, null, "  ")}}
    />
  )
}

export const use${componentName} = createUseComponent(${componentName}, pinNames)

`.trim()
}
