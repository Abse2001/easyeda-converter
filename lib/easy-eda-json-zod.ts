import { z } from "zod"

export const SzlcscSchema = z.object({
  id: z.number(),
  number: z.string(),
  step: z.number(),
  min: z.number(),
  price: z.number(),
  stock: z.number(),
  url: z.string().url(),
  image: z.string().optional(),
})

export const LcscSchema = z.object({
  id: z.number(),
  number: z.string(),
  step: z.number(),
  min: z.number(),
  price: z.number(),
  stock: z.number(),
  url: z.string().url(),
})

export const OwnerSchema = z.object({
  uuid: z.string(),
  username: z.string(),
  nickname: z.string(),
  avatar: z.string(),
})

export const HeadSchema = z.object({
  docType: z.string(),
  editorVersion: z.string(),
  c_para: z.record(z.string(), z.string()),
  x: z.number(),
  y: z.number(),
  puuid: z.string(),
  uuid: z.string(),
  utime: z.number(),
  importFlag: z.number(),
  c_spiceCmd: z.null(),
  hasIdFlag: z.boolean(),
})

export const BBoxSchema = z.object({
  x: z.number(),
  y: z.number(),
  width: z.number(),
  height: z.number(),
})

export const DataStrSchema = z.object({
  head: HeadSchema,
  canvas: z.string(),
  shape: z.array(z.string()),
  BBox: BBoxSchema,
  colors: z.array(z.unknown()),
})

export const PackageDetailSchema = z.object({
  uuid: z.string(),
  title: z.string(),
  docType: z.number(),
  updateTime: z.number(),
  owner: OwnerSchema,
  datastrid: z.string(),
  writable: z.boolean(),
  dataStr: z.object({
    head: HeadSchema,
    canvas: z.string(),
    shape: z.array(z.string()),
    layers: z.array(z.string()),
    objects: z.array(z.string()),
    BBox: BBoxSchema,
    netColors: z.array(z.unknown()),
  }),
})

export const EasyEdaJsonSchema = z.object({
  uuid: z.string(),
  title: z.string(),
  description: z.string(),
  docType: z.number(),
  type: z.number(),
  szlcsc: SzlcscSchema,
  lcsc: LcscSchema,
  owner: OwnerSchema,
  tags: z.array(z.string()),
  updateTime: z.number(),
  updated_at: z.string(),
  dataStr: DataStrSchema,
  verify: z.boolean(),
  SMT: z.boolean(),
  datastrid: z.string(),
  jlcOnSale: z.number(),
  writable: z.boolean(),
  isFavorite: z.boolean(),
  packageDetail: PackageDetailSchema,
})

export type EasyEdaJson = z.infer<typeof EasyEdaJsonSchema>
