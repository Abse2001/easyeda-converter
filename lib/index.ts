export * from "./convert-easyeda-json-to-tscircuit-soup-json"
export * from "./fetch-easyeda-json"
export { convertRawEasyEdaToTs } from "./convert-to-typescript-component"
