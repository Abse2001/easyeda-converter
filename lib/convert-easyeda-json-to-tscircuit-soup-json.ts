import {
  PadSchema,
  TrackSchema,
  ArcSchema,
  SVGNodeSchema,
  HoleSchema,
} from "./schemas/package-detail-shape-schema"
import { z } from "zod"
import type { BetterEasyEdaJson } from "./schemas/easy-eda-json-schema"
import type {
  AnySoupElement,
  PCBSMTPad,
  PcbSilkscreenPath,
  PCBPlatedHole,
  PCBPlatedHoleInput,
  PcbComponentInput,
} from "circuit-json"
import {
  any_source_component,
  pcb_smtpad,
  pcb_silkscreen_path,
  pcb_plated_hole,
  pcb_hole,
} from "circuit-json"
import * as Soup from "circuit-json"
import { generateArcFromSweep, generateArcPathWithMid } from "./math/arc-utils"
import { transformPCBElements } from "@tscircuit/soup-util"
import { scale, translate } from "transformation-matrix"
import { computeCenterOffset } from "./compute-center-offset"
import { mm } from "@tscircuit/mm"

const handleSilkscreenPath = (
  track: z.infer<typeof TrackSchema>,
  index: number,
) => {
  return pcb_silkscreen_path.parse({
    type: "pcb_silkscreen_path",
    pcb_silkscreen_path_id: `pcb_silkscreen_path_${index + 1}`,
    pcb_component_id: "pcb_component_1",
    layer: "top", // Assuming all silkscreen is on top layer
    route: track.points.map((point) => ({ x: point.x, y: point.y })),
    stroke_width: track.width,
  })
}

const handleSilkscreenArc = (arc: z.infer<typeof ArcSchema>, index: number) => {
  const arcPath = generateArcFromSweep(
    arc.start.x,
    arc.start.y,
    arc.end.x,
    arc.end.y,
    arc.radiusX,
    arc.largeArc,
    arc.sweepDirection === "CW",
  )

  return pcb_silkscreen_path.parse({
    type: "pcb_silkscreen_path",
    pcb_silkscreen_path_id: `pcb_silkscreen_arc_${index + 1}`,
    pcb_component_id: "pcb_component_1",
    layer: "top", // Assuming all silkscreen is on top layer
    route: arcPath,
    stroke_width: arc.width,
  })
}

const handleHole = (hole: z.infer<typeof HoleSchema>, index: number) => {
  return pcb_hole.parse({
    type: "pcb_hole",
    x: hole.center.x,
    y: hole.center.y,
    hole_diameter: hole.radius * 2,
    hole_shape: "circle",
    pcb_hole_id: `pcb_hole_${index + 1}`,
  } as Soup.PcbHole)
}

interface Options {
  useModelCdn?: boolean
  shouldRecenter?: boolean
}

export const convertEasyEdaJsonToCircuitJson = (
  easyEdaJson: BetterEasyEdaJson,
  { useModelCdn, shouldRecenter = true }: Options = {},
): AnySoupElement[] => {
  const soupElements: AnySoupElement[] = []
  const centerOffset = computeCenterOffset(easyEdaJson)

  // Add source component
  const source_component = any_source_component.parse({
    type: "source_component",
    source_component_id: "source_component_1",
    name: "U1",
    ftype: "simple_bug",
  })

  const pcb_component = Soup.pcb_component.parse({
    type: "pcb_component",
    pcb_component_id: "pcb_component_1",
    source_component_id: "source_component_1",
    name: "U1",
    ftype: "simple_bug",
    width: 0, // TODO compute width
    height: 0, // TODO compute height
    rotation: 0,
    center: { x: centerOffset.x, y: centerOffset.y },
    layer: "top",
  } as PcbComponentInput)

  soupElements.push(source_component, pcb_component)

  // Add source ports and pcb_smtpads
  easyEdaJson.packageDetail.dataStr.shape
    .filter((shape): shape is z.infer<typeof PadSchema> => shape.type === "PAD")
    .forEach((pad, index) => {
      const portNumber = pad.number.toString()

      // Add source port
      soupElements.push({
        type: "source_port",
        source_port_id: `source_port_${index + 1}`,
        source_component_id: "source_component_1",
        name: portNumber,
      })

      if (pad.holeRadius !== undefined && mm(pad.holeRadius) !== 0) {
        // Add pcb_plated_hole
        soupElements.push(
          pcb_plated_hole.parse({
            type: "pcb_plated_hole",
            pcb_plated_hole_id: `pcb_plated_hole_${index + 1}`,
            shape: "circle",
            x: mm(pad.center.x),
            y: mm(pad.center.y),
            hole_diameter: mm(pad.holeRadius) * 2,
            outer_diameter: mm(pad.width),
            radius: mm(pad.holeRadius),
            port_hints: [portNumber],
            pcb_component_id: "pcb_component_1",
            pcb_port_id: `pcb_port_${index + 1}`,
            layers: ["top"],
          } as PCBPlatedHoleInput),
        )
      } else {
        // Add pcb_smtpad
        let soupShape: PCBSMTPad["shape"] | undefined
        if (pad.shape === "RECT") {
          soupShape = "rect"
        } else if (pad.shape === "ELLIPSE") {
          // This is just a bug
          soupShape = "rect"
        } else if (pad.shape === "OVAL") {
          // OVAL is often a rect, especially when holeRadius is 0
          soupShape = "rect"
        }
        if (!soupShape) {
          throw new Error(`unknown pad.shape: "${pad.shape}"`)
        }
        soupElements.push(
          pcb_smtpad.parse({
            type: "pcb_smtpad",
            pcb_smtpad_id: `pcb_smtpad_${index + 1}`,
            shape: soupShape,
            x: mm(pad.center.x),
            y: mm(pad.center.y),
            ...(soupShape === "rect"
              ? { width: mm(pad.width), height: mm(pad.height) }
              : { radius: Math.min(mm(pad.width), mm(pad.height)) / 2 }),
            layer: "top",
            port_hints: [portNumber],
            pcb_component_id: "pcb_component_1",
            pcb_port_id: `pcb_port_${index + 1}`,
          } as PCBSMTPad),
        )
      }
    })

  // Add holes
  easyEdaJson.packageDetail.dataStr.shape
    .filter(
      (shape): shape is z.infer<typeof HoleSchema> => shape.type === "HOLE",
    )
    .forEach((h, index) => {
      soupElements.push(handleHole(h, index))
    })

  // Add silkscreen paths and arcs
  easyEdaJson.packageDetail.dataStr.shape.forEach((shape, index) => {
    if (shape.type === "TRACK") {
      soupElements.push(handleSilkscreenPath(shape, index))
    } else if (shape.type === "ARC") {
      soupElements.push(handleSilkscreenArc(shape, index))
    }
  })

  // easyeda uses a flipped Y axis ( tscircuit = y+ is up, easyeda = y- is up )
  transformPCBElements(soupElements, scale(1, -1))

  // TODO Change pcb_component width & height

  // TODO compute pcb center based on all elements and transform elements such
  // that the center is (0,0)

  // Add 3d component
  const svgNode = easyEdaJson.packageDetail.dataStr.shape.find(
    (a): a is z.infer<typeof SVGNodeSchema> =>
      Boolean(a.type === "SVGNODE" && a.svgData.attrs?.uuid),
  )
  const objFileUuid = svgNode?.svgData?.attrs?.uuid

  const objFileUrl = objFileUuid
    ? useModelCdn
      ? `https://modelcdn.tscircuit.com/easyeda_models/download?uuid=${objFileUuid}&pn=${easyEdaJson.lcsc.number}`
      : `https://modules.easyeda.com/3dmodel/${objFileUuid}`
    : undefined

  if (objFileUrl !== undefined) {
    const [rx, ry, rz] = (svgNode?.svgData.attrs?.c_rotation ?? "0,0,0")
      .split(",")
      .map(Number)
    soupElements.push(
      Soup.cad_component.parse({
        type: "cad_component",
        cad_component_id: "cad_component_1",
        source_component_id: "source_component_1",
        pcb_component_id: "pcb_component_1",
        position: { x: 0, y: 0, z: 0 },
        rotation: { x: rx, y: ry, z: rz },
        model_obj_url: objFileUrl,
      } as Soup.CadComponentInput),
    )
  }

  if (shouldRecenter) {
    transformPCBElements(
      soupElements,
      translate(-centerOffset.x, centerOffset.y),
    )
  }

  return soupElements
}

/** @deprecated Use `convertEasyEdaJsonToCircuitJson` instead. */
export const convertEasyEdaJsonToTscircuitSoupJson =
  convertEasyEdaJsonToCircuitJson
