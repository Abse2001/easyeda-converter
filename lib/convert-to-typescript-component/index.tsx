import type { AnyCircuitElement } from "circuit-json"
import {
  EasyEdaJsonSchema,
  type BetterEasyEdaJson,
} from "lib/schemas/easy-eda-json-schema"
import { su } from "@tscircuit/soup-util"
import { soupTypescriptComponentTemplate } from "./soup-typescript-component-template"
import {
  convertEasyEdaJsonToCircuitJson,
  convertEasyEdaJsonToTscircuitSoupJson,
} from "lib/convert-easyeda-json-to-tscircuit-soup-json"
import { normalizeManufacturerPartNumber } from "lib/utils/normalize-manufacturer-part-number"

export const convertRawEasyToTsx = async (rawEasy: any) => {
  const betterEasy = EasyEdaJsonSchema.parse(rawEasy)
  const result = await convertBetterEasyToTsx({
    betterEasy,
  })
  return result
}

export const convertBetterEasyToTsx = async ({
  betterEasy,
}: {
  betterEasy: BetterEasyEdaJson
}): Promise<string> => {
  const circuitJson = convertEasyEdaJsonToCircuitJson(betterEasy, {
    useModelCdn: true,
    shouldRecenter: true,
  })
  const rawPn = betterEasy.dataStr.head.c_para["Manufacturer Part"]
  const pn = normalizeManufacturerPartNumber(rawPn)
  const [cad_component] = su(circuitJson).cad_component.list()

  // Derive schPinArrangement from easyeda json
  const pins = betterEasy.dataStr.shape.filter((shape) => shape.type === "PIN")

  const hasStringPinNumbers = pins.some(
    (pin) => typeof pin.pinNumber === "string",
  )
  let modifiedPins = pins
  if (hasStringPinNumbers) {
    modifiedPins = pins.map((pin, idx) => {
      const originalPinNumber = pin.pinNumber.toString()

      const newPinNumber = idx + 1

      // If label is different from original pin number, create array with both pin number and label
      // const newLabel =
      //   originalPinNumber === pin.label
      //     ? [pin.label]
      //     : [originalPinNumber, pin.label]

      return {
        ...pin,
        pinNumber: newPinNumber,
        label: pin.label,
      }
    })
  }

  const leftPins = modifiedPins.filter((pin) => pin.rotation === 180)
  const rightPins = modifiedPins.filter((pin) => pin.rotation === 0)

  const schPinArrangement = {
    leftSide: {
      direction: "top-to-bottom" as const,
      pins: leftPins.map((pin) => Number(pin.pinNumber)),
    },
    rightSide: {
      direction: "bottom-to-top" as const,
      pins: rightPins.map((pin) => Number(pin.pinNumber)).reverse(),
    },
  }

  // Derive pinLabels from easyeda json
  const pinLabels: Record<string, string> = {}
  for (const pin of modifiedPins) {
    const isPinLabelNumeric = /^\d+$/.test(pin.label)
    const label = isPinLabelNumeric ? `pin${pin.label}` : pin.label

    pinLabels[pin.pinNumber] = label
  }

  let modelObjUrl: string | undefined
  if (cad_component.model_obj_url) {
    const isValidUrl = await checkModelObjUrlValidity(
      cad_component.model_obj_url,
    )
    if (isValidUrl) {
      modelObjUrl = cad_component.model_obj_url
    }
  }

  const supplierPartNumbers: Record<string, string[]> = {
    lcsc: [betterEasy.lcsc.number],
  }

  return soupTypescriptComponentTemplate({
    componentName: pn,
    pinLabels,
    schPinArrangement,
    objUrl: modelObjUrl,
    circuitJson,
    supplierPartNumbers,
  })
}

const checkModelObjUrlValidity = async (url: string): Promise<boolean> => {
  try {
    const response = await fetch(url, { method: "HEAD" })
    return response.status === 200
  } catch (error) {
    console.error(`Error checking model object URL ${url}:`, error)
    return false
  }
}
