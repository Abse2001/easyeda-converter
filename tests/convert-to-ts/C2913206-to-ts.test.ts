import { it, expect } from "bun:test"
import chipRawEasy from "../assets/C2913206.raweasy.json"
import { convertBetterEasyToTsx } from "lib/convert-to-typescript-component"
import { EasyEdaJsonSchema } from "lib/schemas/easy-eda-json-schema"

it("should convert C2913206 into typescript file", async () => {
  const betterEasy = EasyEdaJsonSchema.parse(chipRawEasy)
  const result = await convertBetterEasyToTsx({
    betterEasy,
  })

  expect(result).not.toContain("milmm")
  expect(result).not.toContain("NaNmm")

  const pinLabelString = result.match(/const pinLabels = \{[\s\S]*?\}/gm)?.[0]
  console.log(pinLabelString)
  expect(pinLabelString).toMatchSnapshot()
  // Add more specific assertions here based on the component
})
