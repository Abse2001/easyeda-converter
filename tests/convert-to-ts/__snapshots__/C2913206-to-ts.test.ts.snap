// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should convert C2913206 into typescript file 1`] = `
"const pinLabels = {
  "1": "pin1",
  "2": "pin2",
  "3": "pin3",
  "4": "pin4",
  "5": "pin5",
  "6": "pin6",
  "7": "pin7",
  "8": "pin8",
  "9": "pin9",
  "10": "pin10",
  "11": "pin11",
  "12": "pin12",
  "13": "pin13",
  "14": "pin14",
  "15": "pin15",
  "16": "pin16",
  "17": "pin17",
  "18": "pin18",
  "19": "pin19",
  "20": "pin20",
  "21": "pin21",
  "22": "pin22",
  "23": "pin23",
  "24": "pin24",
  "25": "pin25",
  "26": "pin26",
  "27": "pin27",
  "28": "pin28",
  "29": "pin29",
  "30": "pin30",
  "31": "pin31",
  "32": "pin32",
  "33": "pin33",
  "34": "pin34",
  "35": "pin35",
  "36": "pin36",
  "37": "pin37",
  "38": "pin38",
  "39": "pin39",
  "40": "pin40",
  "41": "pin41",
  "42": "pin42",
  "43": "pin43",
  "44": "pin44",
  "45": "pin45",
  "46": "pin46",
  "47": "pin47",
  "48": "pin48",
  "49": "pin49",
  "50": "pin50",
  "51": "pin51",
  "52": "pin52",
  "53": "pin53",
  "54": "pin54",
  "55": "pin55",
  "56": "pin56",
  "57": "pin57",
  "58": "pin58",
  "59": "pin59",
  "60": "pin60",
  "61": "pin61"
}"
`;
